src/Streamly/Internal/Data/Parse.hs

```diff
@@ -252,3 +277,101 @@ endBefore = undefined
 --
 finishBy :: Parse m a x -> Fold m a y -> Parse m a (x, y)
 finishBy = undefined
+
+-------------------------------------------------------------------------------
+-- Zipping Parses
+-------------------------------------------------------------------------------
+--
+
+-- | State of the pair of parsers in a zip composition
+-- Note: strictness annotation is important for fusing the constructors
+{-# ANN type ParState Fuse #-}
+data ParState sL sR a =
+    ParPair !([a], Either sL sL, [a], [a]) !([a], Either sR sR, [a], [a])
+
+-- XXX we can write a much simpler zipWith if the types do not have Back or
+-- Hold. Perhaps we can write a simpler partial function assuming that and
+-- statically assert the assumptions with the help of the compiler.
+--
+-- zipWithMin -- end when either stream ends
+-- zipWithFst -- end when the first stream ends
+--
+-- XXX requires -funfolding-use-threshold=200 to fuse
+--
+-- | @zipWith f p1 p2@ distributes its input to both @p1@ and @p2@ and combines
+-- their output using @f@.
+--
+-- /Internal/
+--
+{-# INLINE zipWith #-}
+zipWith :: Monad m
+    => (a -> b -> c) -> Parse m x a -> Parse m x b -> Parse m x c
+zipWith zf (Parse stepL initialL extractL) (Parse stepR initialR extractR) =
+    Parse step initial extract
+
+    where
+
+    {-# INLINE_LATE initial #-}
+    initial = do
+        sL <- initialL
+        sR <- initialR
+        return $ ParPair ([], Right sL, [], []) ([], Right sR, [], [])
+
+    {-# INLINE_LATE step #-}
+    step (ParPair (bufL, Right sL, inpL1, inpL2)
+                  (bufR, Right sR, inpR1, inpR2)) x = do
```

Comment:
What is `inpL1` and `inpL2` here?

Reply:
Won't `inpR2` and `inpL2` always be `[]`?

Reply:
I believe you are using these for efficiency? Similar to how d-list is implemented for efficient `snoc` and `reverse`?

Reply:
Cool, Cool!